plonky2/src/plonk/circuit_builder.rs

```diff
@@ -695,12 +740,26 @@ impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
             constants_sigmas_cap: constants_sigmas_cap.clone(),
         };
 
+        // Map between gates where not all generators are used and the gate's number of used generators.
+        let incomplete_gates = self
+            .current_slots
+            .values()
+            .flat_map(|current_slot| current_slot.current_slot.values().copied())
+            .collect::<HashMap<_, _>>();
+
         // Add gate generators.
         self.add_generators(
             self.gate_instances
                 .iter()
                 .enumerate()
-                .flat_map(|(index, gate)| gate.gate_ref.0.generators(index, &gate.constants))
+                .flat_map(|(index, gate)| {
+                    let mut gens = gate.gate_ref.0.generators(index, &gate.constants);
+                    // Remove unused generators, if any.
```

Comment:
This is a nice solution, my only concern is it seems like "batch" gates now must have one generator per slot. I think it's okay, but to avoid any surprises it might be good to document the requirement (in `generators()`?), and/or maybe assert `gens.len() == num_ops`.

Or to make it even safer, maybe we could change `generators()` to return a single generator for a particular slot index. That would be a major change though so probably not now.

Reply:
So the default `num_ops()` just returns the number of generators:
```rust
fn num_ops(&self) -> usize {
    self.generators(0, &vec![F::ZERO; self.num_constants()])
        .len()
}
```
I agree it would be nice to disallow overwriting this function. Maybe just adding a comment: "Don't overwrite!" would be ok?

Reply:
> Or to make it even safer, maybe we could change `generators()` to return a single generator for a particular slot index. That would be a major change though so probably not now.

Yeah I've thought about doing this, I think that would also work but I kind of like the current solution where `generators()` is the only source of truth and dictates the number of operations. I think that makes it really easy to write non-batchable gate without having to think about any batch-related logic.

Reply:
Oh I see. I wouldn't really worry about `num_ops` being overridden, but I think the other risk is still there, e.g. someone could write a gate that supported `n` operations but used `2n` generators.

I feel like the API change would be safest, though a comment that gates should return one generator per slot that they actually support could also work.

Reply:
Hmm is there a use case for a gate with more than one generator per slot?

Reply:
I don't think it's something important we need to support, but I think there's a decent chance someone could attempt to write it that way, e.g. if they were trying to reuse some existing generators and compose them together.